gelfling.js

```diff
@@ -18,17 +19,73 @@ function Gelfling(host, port, options) {
   this.defaults = options.defaults || {}
   this.errHandler = options.errHandler || console.error
   this.keepAlive = options.keepAlive
+  this.tcp = !!options.tcp
 }
 
-Gelfling.prototype.send = function(data, callback) {
-  if (callback == null) callback = function() {}
+function callOnce(fn) {
+  if (fn == null) return function() {}
+  var called = false
+  return function() {
+    if (! called) {
+      called = true
+      return fn.apply(this, arguments)
+    }
+  }
+}
+
+Gelfling.prototype._connectTcpClient = function(callback) {
+  if (this.tcpConnected || this.tcpConnecting) return
+  var that = this
+  callback = callOnce(callback)
+  this.tcpConnecting = true
+  this.tcpClient = net.connect(this.port, this.host)
+  this.tcpClient
+    .on('connect', function() {
+      that.tcpConnecting = false
+      that.tcpConnected = true
+      callback()
+    })
+    .on('error', function(err) {
+      that.tcpConnecting = false
+      that.tcpConnected = false
+      callback(err)
+    })
+    .on('close', function() {
+      that.tcpConnecting = false
+      that.tcpConnected = false
+    })
+    .on('end', function() {
+      that.tcpConnecting = false
+      that.tcpConnected = false
+    })
+}
+
+Gelfling.prototype._sendTcp = function(data, callback) {
+  var that = this
+  var retrySend = function(err) {
+    if (err) return callback(err)
+    that._sendTcp(data, callback)
+  }
+  if (this.tcpConnected) {
+    // No GZIP for TCP. See: 
+    //   https://github.com/Moocar/logback-gelf#tcp
+    //   https://github.com/Graylog2/graylog2-server/issues/127
+    //   https://github.com/t0xa/gelfj/pull/61
+    this.tcpClient.write(JSON.stringify(this.convert(data)) + '\0', callback)
+  } else if (this.tcpConnecting)
+    setTimeout(retrySend, 10)
```

Comment:
This may result in race conditions and your data will be sent out of order. Eg:

1. say `tcpConnecting` is true
2. `_sendTcp` is called with some data, so `setTimeout` will retry in 10ms
3. in that time, `tcpConnecting` is set to true
4. then let's say `_sendTcp` is called with new data – that will now be sent first
5. then when the `setTimeout` triggers, the first packet of data will be sent second

One way to avoid this is to create a queue – it could just be an array of bound callbacks – and then call each of these on connect.

I guess the other thing to do is just ignore it – and live with the fact that out of order messages may occur :smile_cat: